File: solutions/csharp/circular-buffer/1/CircularBuffer.cs
public class CircularBuffer<T>
{
    private readonly T[] _buffer;
    private readonly int _capacity;
    private int _readPtr;
    private int _writePtr;

    private int Count => _writePtr - _readPtr;
    private bool IsEmpty => Count == 0;
    private bool IsFull => Count == _capacity;

    public CircularBuffer(int capacity)
    {
        _capacity = capacity;
        _buffer = new T[capacity];
        _writePtr = _readPtr = 0;
    }

    public T Read()
    {
        Require(!IsEmpty);

        return _buffer[_readPtr++ % _capacity];
    }

    public void Write(T value)
    {
        Require(!IsFull);

        _buffer[_writePtr++ % _capacity] = value;
    }

    public void Overwrite(T value)
    {
        if (IsFull)
        {
            DiscardOldestItem();
        }

        _buffer[_writePtr++ % _capacity] = value;
    }

    public void Clear()
    {
        _readPtr = _writePtr;
    }

    private void DiscardOldestItem()
    {
        _readPtr++;
    }

    private void Require(bool condition)
    {
        if (!condition)
        {
            throw new InvalidOperationException();
        }
    }
}

File: solutions/scala/circular-buffer/1/src/main/scala/CircularBuffer.scala
class EmptyBufferException() extends Exception {}

class FullBufferException() extends Exception {}

class CircularBuffer(capacity: Int) {
  private val values = Array.fill(capacity)(0)
  private var readIndex = 0
  private var writeIndex = 0
  private var count = 0

  def write(value: Int): Unit =
    if isFull then throw FullBufferException()
    else writeUnchecked(value)

  def read(): Int =
    if isEmpty then throw EmptyBufferException()
    else
      val value = values(readIndex)
      readIndex = next(readIndex)
      count -= 1
      value

  def overwrite(value: Int): Unit =
    if isFull then
      values(writeIndex) = value
      writeIndex = next(writeIndex)
      readIndex = writeIndex
    else writeUnchecked(value)

  def clear(): Unit =
    readIndex = 0
    writeIndex = 0
    count = 0

  private def writeUnchecked(value: Int): Unit =
    values(writeIndex) = value
    writeIndex = next(writeIndex)
    count += 1

  private def isFull: Boolean =
    count == capacity

  private def isEmpty: Boolean =
    count == 0

  private def next(index: Int): Int =
    (index + 1) % capacity
}
